error: field 'ts' has incomplete type

Hi,

I'm trying to compile netmap against the Linux kernel v5.6 (v5.5 seems to build fine). I'm running into issues regarding `struct timeval` I know that there have been some on-going changes in the kernel related to this type so it's probably got something to do with that.

The thing is bsd_glue.h is picking up linux/time.h but somehow it's not getting the definition. I suspect something is being clever and stopping the use of uapi headers in kernel modules.

<details>
<summary>Error output</summary>

```
********************************** WARNING **********************************
*** Unrecognized option: ,octeon
*****************************************************************************
**********************************  NOTE   **********************************
*** Running some preliminary tests to customize the build enviroment.
*****************************************************************************
**********************************  NOTE   **********************************
*** Now running compile tests to adapt the code to your
*** kernel version. Please wait.
*****************************************************************************
kernel directory            /src/staging/lib/modules/5.6.0-at1/build
                            [/src/linux-build]
kernel sources              /src/staging/lib/modules/5.6.0-at1/build/source
                            [/src/linux]
linux version               50600  [5.6]
module file                 netmap.ko

subsystems                  null generic pipe vale
apps                        vale-ctl nmreplay tlem lb bridge pkt-gen
native drivers              

make -C /src/staging/lib/modules/5.6.0-at1/build M=/src/netmap/LINUX EXTRA_CFLAGS='-I/src/netmap/LINUX -I/src/netmap/LINUX -I/src/netmap/LINUX/../sys -I/src/netmap/LINUX/../sys/dev -DCONFIG_NETMAP -Wno-unused-but-set-variable -g -DCONFIG_NETMAP_NULL -DCONFIG_NETMAP_GENERIC -DCONFIG_NETMAP_PIPE -DCONFIG_NETMAP_VALE' ARCH=mips CROSS_COMPILE=mips64-unknown-linux-gnu- TARGET=mips64-unknown-linux-gnu KCPPFLAGS=-DATL_CHANGE KCFLAGS=-Werror -Wno-error=cpp TARGET_GCC=mips64-unknown-linux-gnu-gcc modules CONFIG_NETMAP=m  O_DRIVERS="" NETMAP_DRIVER_SUFFIX=
make[1]: Entering directory '/src/linux-build'
  CC [M]  /src/netmap/LINUX/netmap_mem2.o
In file included from /src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c:67:
/src/netmap/LINUX/../sys/net/netmap.h:298:17: error: field 'ts' has incomplete type
  struct timeval ts;  /* (k) time of last *sync() */
                 ^~
In file included from /src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c:68:
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c: In function 'netmap_init_obj_allocator_bitmap':
/src/netmap/LINUX/../sys/dev/netmap/netmap_kern.h:270:18: error: storage size of '__xxts' isn't known
   struct timeval __xxts;    \
                  ^~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c:403:4: note: in expansion of macro 'nm_prerr'
    nm_prerr("Unable to create bitmap (%d entries) for allocator '%s'", (int)n,
    ^~~~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_kern.h:270:18: error: unused variable '__xxts' [-Werror=unused-variable]
   struct timeval __xxts;    \
                  ^~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c:403:4: note: in expansion of macro 'nm_prerr'
    nm_prerr("Unable to create bitmap (%d entries) for allocator '%s'", (int)n,
    ^~~~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_kern.h:261:18: error: storage size of '__xxts' isn't known
   struct timeval __xxts;    \
                  ^~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c:421:5: note: in expansion of macro 'nm_prinf'
     nm_prinf("skipping %s %d", p->name, j);
     ^~~~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_kern.h:261:18: error: unused variable '__xxts' [-Werror=unused-variable]
   struct timeval __xxts;    \
                  ^~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c:421:5: note: in expansion of macro 'nm_prinf'
     nm_prinf("skipping %s %d", p->name, j);
     ^~~~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_kern.h:261:18: error: storage size of '__xxts' isn't known
   struct timeval __xxts;    \
                  ^~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c:429:3: note: in expansion of macro 'nm_prinf'
   nm_prinf("%s free %u", p->name, p->objfree);
   ^~~~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_kern.h:261:18: error: unused variable '__xxts' [-Werror=unused-variable]
   struct timeval __xxts;    \
                  ^~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c:429:3: note: in expansion of macro 'nm_prinf'
   nm_prinf("%s free %u", p->name, p->objfree);
   ^~~~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_kern.h:270:18: error: storage size of '__xxts' isn't known
   struct timeval __xxts;    \
                  ^~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c:432:4: note: in expansion of macro 'nm_prerr'
    nm_prerr("%s: no objects available", p->name);
    ^~~~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_kern.h:270:18: error: unused variable '__xxts' [-Werror=unused-variable]
   struct timeval __xxts;    \
                  ^~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c:432:4: note: in expansion of macro 'nm_prerr'
    nm_prerr("%s: no objects available", p->name);
    ^~~~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c: In function 'netmap_mem_init_bitmaps':
/src/netmap/LINUX/../sys/dev/netmap/netmap_kern.h:270:18: error: storage size of '__xxts' isn't known
   struct timeval __xxts;    \
                  ^~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c:456:3: note: in expansion of macro 'nm_prerr'
   nm_prerr("%s: not enough buffers", nmd->pools[NETMAP_BUF_POOL].name);
   ^~~~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_kern.h:270:18: error: unused variable '__xxts' [-Werror=unused-variable]
   struct timeval __xxts;    \
                  ^~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c:456:3: note: in expansion of macro 'nm_prerr'
   nm_prerr("%s: not enough buffers", nmd->pools[NETMAP_BUF_POOL].name);
   ^~~~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c: In function 'nm_mem_assign_group':
/src/netmap/LINUX/../sys/dev/netmap/netmap_kern.h:261:18: error: storage size of '__xxts' isn't known
   struct timeval __xxts;    \
                  ^~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c:733:3: note: in expansion of macro 'nm_prinf'
   nm_prinf("iommu_group %d", id);
   ^~~~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_kern.h:261:18: error: unused variable '__xxts' [-Werror=unused-variable]
   struct timeval __xxts;    \
                  ^~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c:733:3: note: in expansion of macro 'nm_prinf'
   nm_prinf("iommu_group %d", id);
   ^~~~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_kern.h:270:18: error: storage size of '__xxts' isn't known
   struct timeval __xxts;    \
                  ^~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c:742:4: note: in expansion of macro 'nm_prerr'
    nm_prerr("iommu group mismatch: %u vs %u",
    ^~~~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_kern.h:270:18: error: unused variable '__xxts' [-Werror=unused-variable]
   struct timeval __xxts;    \
                  ^~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c:742:4: note: in expansion of macro 'nm_prerr'
    nm_prerr("iommu group mismatch: %u vs %u",
    ^~~~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c: In function 'netmap_mem2_ofstophys':
/src/netmap/LINUX/../sys/dev/netmap/netmap_kern.h:270:18: error: storage size of '__xxts' isn't known
   struct timeval __xxts;    \
                  ^~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c:821:2: note: in expansion of macro 'nm_prerr'
  nm_prerr("invalid ofs 0x%x out of 0x%zx 0x%zx 0x%zx", (u_int)o,
  ^~~~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_kern.h:270:18: error: unused variable '__xxts' [-Werror=unused-variable]
   struct timeval __xxts;    \
                  ^~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c:821:2: note: in expansion of macro 'nm_prerr'
  nm_prerr("invalid ofs 0x%x out of 0x%zx 0x%zx 0x%zx", (u_int)o,
  ^~~~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c: In function 'netmap_obj_offset':
/src/netmap/LINUX/../sys/dev/netmap/netmap_kern.h:270:18: error: storage size of '__xxts' isn't known
   struct timeval __xxts;    \
                  ^~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c:987:2: note: in expansion of macro 'nm_prerr'
  nm_prerr("address %p is not contained inside any cluster (%s)",
  ^~~~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_kern.h:270:18: error: unused variable '__xxts' [-Werror=unused-variable]
   struct timeval __xxts;    \
                  ^~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c:987:2: note: in expansion of macro 'nm_prerr'
  nm_prerr("address %p is not contained inside any cluster (%s)",
  ^~~~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c: In function 'netmap_obj_malloc':
/src/netmap/LINUX/../sys/dev/netmap/netmap_kern.h:270:18: error: storage size of '__xxts' isn't known
   struct timeval __xxts;    \
                  ^~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c:1018:3: note: in expansion of macro 'nm_prerr'
   nm_prerr("%s request size %d too large", p->name, len);
   ^~~~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_kern.h:270:18: error: unused variable '__xxts' [-Werror=unused-variable]
   struct timeval __xxts;    \
                  ^~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c:1018:3: note: in expansion of macro 'nm_prerr'
   nm_prerr("%s request size %d too large", p->name, len);
   ^~~~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_kern.h:270:18: error: storage size of '__xxts' isn't known
   struct timeval __xxts;    \
                  ^~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c:1023:3: note: in expansion of macro 'nm_prerr'
   nm_prerr("no more %s objects", p->name);
   ^~~~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_kern.h:270:18: error: unused variable '__xxts' [-Werror=unused-variable]
   struct timeval __xxts;    \
                  ^~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c:1023:3: note: in expansion of macro 'nm_prerr'
   nm_prerr("no more %s objects", p->name);
   ^~~~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c: In function 'netmap_obj_free':
/src/netmap/LINUX/../sys/dev/netmap/netmap_kern.h:270:18: error: storage size of '__xxts' isn't known
   struct timeval __xxts;    \
                  ^~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c:1065:3: note: in expansion of macro 'nm_prerr'
   nm_prerr("invalid index %u, max %u", j, p->objtotal);
   ^~~~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_kern.h:270:18: error: unused variable '__xxts' [-Werror=unused-variable]
   struct timeval __xxts;    \
                  ^~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c:1065:3: note: in expansion of macro 'nm_prerr'
   nm_prerr("invalid index %u, max %u", j, p->objtotal);
   ^~~~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_kern.h:270:18: error: storage size of '__xxts' isn't known
   struct timeval __xxts;    \
                  ^~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c:1071:3: note: in expansion of macro 'nm_prerr'
   nm_prerr("ouch, double free on buffer %d", j);
   ^~~~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_kern.h:270:18: error: unused variable '__xxts' [-Werror=unused-variable]
   struct timeval __xxts;    \
                  ^~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c:1071:3: note: in expansion of macro 'nm_prerr'
   nm_prerr("ouch, double free on buffer %d", j);
   ^~~~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c: In function 'netmap_obj_free_va':
/src/netmap/LINUX/../sys/dev/netmap/netmap_kern.h:270:18: error: storage size of '__xxts' isn't known
   struct timeval __xxts;    \
                  ^~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c:1102:2: note: in expansion of macro 'nm_prerr'
  nm_prerr("address %p is not contained inside any cluster (%s)",
  ^~~~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_kern.h:270:18: error: unused variable '__xxts' [-Werror=unused-variable]
   struct timeval __xxts;    \
                  ^~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c:1102:2: note: in expansion of macro 'nm_prerr'
  nm_prerr("address %p is not contained inside any cluster (%s)",
  ^~~~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c: In function 'netmap_extra_alloc':
/src/netmap/LINUX/../sys/dev/netmap/netmap_kern.h:270:18: error: storage size of '__xxts' isn't known
   struct timeval __xxts;    \
                  ^~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c:1143:4: note: in expansion of macro 'nm_prerr'
    nm_prerr("no more buffers after %d of %d", i, n);
    ^~~~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_kern.h:270:18: error: unused variable '__xxts' [-Werror=unused-variable]
   struct timeval __xxts;    \
                  ^~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c:1143:4: note: in expansion of macro 'nm_prerr'
    nm_prerr("no more buffers after %d of %d", i, n);
    ^~~~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c: In function 'netmap_extra_free':
/src/netmap/LINUX/../sys/dev/netmap/netmap_kern.h:270:18: error: storage size of '__xxts' isn't known
   struct timeval __xxts;    \
                  ^~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c:1174:3: note: in expansion of macro 'nm_prerr'
   nm_prerr("breaking with head %d", head);
   ^~~~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_kern.h:270:18: error: unused variable '__xxts' [-Werror=unused-variable]
   struct timeval __xxts;    \
                  ^~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c:1174:3: note: in expansion of macro 'nm_prerr'
   nm_prerr("breaking with head %d", head);
   ^~~~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_kern.h:261:18: error: storage size of '__xxts' isn't known
   struct timeval __xxts;    \
                  ^~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c:1176:3: note: in expansion of macro 'nm_prinf'
   nm_prinf("freed %d buffers", i);
   ^~~~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_kern.h:261:18: error: unused variable '__xxts' [-Werror=unused-variable]
   struct timeval __xxts;    \
                  ^~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c:1176:3: note: in expansion of macro 'nm_prinf'
   nm_prinf("freed %d buffers", i);
   ^~~~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c: In function 'netmap_new_bufs':
/src/netmap/LINUX/../sys/dev/netmap/netmap_kern.h:270:18: error: storage size of '__xxts' isn't known
   struct timeval __xxts;    \
                  ^~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c:1192:4: note: in expansion of macro 'nm_prerr'
    nm_prerr("no more buffers after %d of %d", i, n);
    ^~~~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_kern.h:270:18: error: unused variable '__xxts' [-Werror=unused-variable]
   struct timeval __xxts;    \
                  ^~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c:1192:4: note: in expansion of macro 'nm_prerr'
    nm_prerr("no more buffers after %d of %d", i, n);
    ^~~~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c: In function 'netmap_free_buf':
/src/netmap/LINUX/../sys/dev/netmap/netmap_kern.h:270:18: error: storage size of '__xxts' isn't known
   struct timeval __xxts;    \
                  ^~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c:1233:3: note: in expansion of macro 'nm_prerr'
   nm_prerr("Cannot free buf#%d: should be in [2, %d[", i, p->objtotal);
   ^~~~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_kern.h:270:18: error: unused variable '__xxts' [-Werror=unused-variable]
   struct timeval __xxts;    \
                  ^~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c:1233:3: note: in expansion of macro 'nm_prerr'
   nm_prerr("Cannot free buf#%d: should be in [2, %d[", i, p->objtotal);
   ^~~~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c: In function 'netmap_config_obj_allocator':
/src/netmap/LINUX/../sys/dev/netmap/netmap_kern.h:270:18: error: storage size of '__xxts' isn't known
   struct timeval __xxts;    \
                  ^~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c:1333:3: note: in expansion of macro 'nm_prerr'
   nm_prerr("unsupported allocation for %d bytes", objsize);
   ^~~~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_kern.h:270:18: error: unused variable '__xxts' [-Werror=unused-variable]
   struct timeval __xxts;    \
                  ^~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c:1333:3: note: in expansion of macro 'nm_prerr'
   nm_prerr("unsupported allocation for %d bytes", objsize);
   ^~~~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_kern.h:261:18: error: storage size of '__xxts' isn't known
   struct timeval __xxts;    \
                  ^~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c:1339:3: note: in expansion of macro 'nm_prinf'
   nm_prinf("aligning object by %d bytes", LINE_ROUND - i);
   ^~~~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_kern.h:261:18: error: unused variable '__xxts' [-Werror=unused-variable]
   struct timeval __xxts;    \
                  ^~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c:1339:3: note: in expansion of macro 'nm_prinf'
   nm_prinf("aligning object by %d bytes", LINE_ROUND - i);
   ^~~~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_kern.h:270:18: error: storage size of '__xxts' isn't known
   struct timeval __xxts;    \
                  ^~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c:1343:3: note: in expansion of macro 'nm_prerr'
   nm_prerr("requested objsize %d out of range [%d, %d]",
   ^~~~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_kern.h:270:18: error: unused variable '__xxts' [-Werror=unused-variable]
   struct timeval __xxts;    \
                  ^~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c:1343:3: note: in expansion of macro 'nm_prerr'
   nm_prerr("requested objsize %d out of range [%d, %d]",
   ^~~~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_kern.h:270:18: error: storage size of '__xxts' isn't known
   struct timeval __xxts;    \
                  ^~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c:1348:3: note: in expansion of macro 'nm_prerr'
   nm_prerr("requested objtotal %d out of range [%d, %d]",
   ^~~~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_kern.h:270:18: error: unused variable '__xxts' [-Werror=unused-variable]
   struct timeval __xxts;    \
                  ^~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c:1348:3: note: in expansion of macro 'nm_prerr'
   nm_prerr("requested objtotal %d out of range [%d, %d]",
   ^~~~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_kern.h:270:18: error: storage size of '__xxts' isn't known
   struct timeval __xxts;    \
                  ^~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c:1370:3: note: in expansion of macro 'nm_prerr'
   nm_prerr("unsupported allocation for %d bytes", objsize);
   ^~~~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_kern.h:270:18: error: unused variable '__xxts' [-Werror=unused-variable]
   struct timeval __xxts;    \
                  ^~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c:1370:3: note: in expansion of macro 'nm_prerr'
   nm_prerr("unsupported allocation for %d bytes", objsize);
   ^~~~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_kern.h:261:18: error: storage size of '__xxts' isn't known
   struct timeval __xxts;    \
                  ^~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c:1376:3: note: in expansion of macro 'nm_prinf'
   nm_prinf("objsize %d clustsize %d objects %d",
   ^~~~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_kern.h:261:18: error: unused variable '__xxts' [-Werror=unused-variable]
   struct timeval __xxts;    \
                  ^~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c:1376:3: note: in expansion of macro 'nm_prinf'
   nm_prinf("objsize %d clustsize %d objects %d",
   ^~~~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c: In function 'netmap_finalize_obj_allocator':
/src/netmap/LINUX/../sys/dev/netmap/netmap_kern.h:270:18: error: storage size of '__xxts' isn't known
   struct timeval __xxts;    \
                  ^~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c:1419:3: note: in expansion of macro 'nm_prerr'
   nm_prerr("Unable to create lookup table for '%s'", p->name);
   ^~~~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_kern.h:270:18: error: unused variable '__xxts' [-Werror=unused-variable]
   struct timeval __xxts;    \
                  ^~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c:1419:3: note: in expansion of macro 'nm_prerr'
   nm_prerr("Unable to create lookup table for '%s'", p->name);
   ^~~~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_kern.h:270:18: error: storage size of '__xxts' isn't known
   struct timeval __xxts;    \
                  ^~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c:1446:4: note: in expansion of macro 'nm_prerr'
    nm_prerr("Unable to create cluster at %d for '%s' allocator",
    ^~~~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_kern.h:270:18: error: unused variable '__xxts' [-Werror=unused-variable]
   struct timeval __xxts;    \
                  ^~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c:1446:4: note: in expansion of macro 'nm_prerr'
    nm_prerr("Unable to create cluster at %d for '%s' allocator",
    ^~~~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_kern.h:261:18: error: storage size of '__xxts' isn't known
   struct timeval __xxts;    \
                  ^~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c:1482:3: note: in expansion of macro 'nm_prinf'
   nm_prinf("Pre-allocated %d clusters (%d/%zuKB) for '%s'",
   ^~~~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_kern.h:261:18: error: unused variable '__xxts' [-Werror=unused-variable]
   struct timeval __xxts;    \
                  ^~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c:1482:3: note: in expansion of macro 'nm_prinf'
   nm_prinf("Pre-allocated %d clusters (%d/%zuKB) for '%s'",
   ^~~~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c: In function 'netmap_mem_reset_all':
/src/netmap/LINUX/../sys/dev/netmap/netmap_kern.h:261:18: error: storage size of '__xxts' isn't known
   struct timeval __xxts;    \
                  ^~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c:1515:3: note: in expansion of macro 'nm_prinf'
   nm_prinf("resetting %p", nmd);
   ^~~~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_kern.h:261:18: error: unused variable '__xxts' [-Werror=unused-variable]
   struct timeval __xxts;    \
                  ^~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c:1515:3: note: in expansion of macro 'nm_prinf'
   nm_prinf("resetting %p", nmd);
   ^~~~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c: In function 'netmap_mem_map':
/src/netmap/LINUX/../sys/dev/netmap/netmap_kern.h:270:18: error: storage size of '__xxts' isn't known
   struct timeval __xxts;    \
                  ^~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c:1591:3: note: in expansion of macro 'nm_prerr'
   nm_prerr("Failed to allocate physical lut for %s", na->name);
   ^~~~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_kern.h:270:18: error: unused variable '__xxts' [-Werror=unused-variable]
   struct timeval __xxts;    \
                  ^~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c:1591:3: note: in expansion of macro 'nm_prerr'
   nm_prerr("Failed to allocate physical lut for %s", na->name);
   ^~~~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_kern.h:270:18: error: storage size of '__xxts' isn't known
   struct timeval __xxts;    \
                  ^~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c:1608:4: note: in expansion of macro 'nm_prerr'
    nm_prerr("Failed to map cluster #%d from the %s pool", i, p->name);
    ^~~~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_kern.h:270:18: error: unused variable '__xxts' [-Werror=unused-variable]
   struct timeval __xxts;    \
                  ^~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c:1608:4: note: in expansion of macro 'nm_prerr'
    nm_prerr("Failed to map cluster #%d from the %s pool", i, p->name);
    ^~~~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c: In function 'netmap_mem_finalize_all':
/src/netmap/LINUX/../sys/dev/netmap/netmap_kern.h:261:18: error: storage size of '__xxts' isn't known
   struct timeval __xxts;    \
                  ^~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c:1646:3: note: in expansion of macro 'nm_prinf'
   nm_prinf("interfaces %zd KB, rings %zd KB, buffers %zd MB",
   ^~~~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_kern.h:261:18: error: unused variable '__xxts' [-Werror=unused-variable]
   struct timeval __xxts;    \
                  ^~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c:1646:3: note: in expansion of macro 'nm_prinf'
   nm_prinf("interfaces %zd KB, rings %zd KB, buffers %zd MB",
   ^~~~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_kern.h:261:18: error: storage size of '__xxts' isn't known
   struct timeval __xxts;    \
                  ^~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c:1652:3: note: in expansion of macro 'nm_prinf'
   nm_prinf("Free buffers: %d", nmd->pools[NETMAP_BUF_POOL].objfree);
   ^~~~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_kern.h:261:18: error: unused variable '__xxts' [-Werror=unused-variable]
   struct timeval __xxts;    \
                  ^~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c:1652:3: note: in expansion of macro 'nm_prinf'
   nm_prinf("Free buffers: %d", nmd->pools[NETMAP_BUF_POOL].objfree);
   ^~~~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c: In function 'netmap_mem_private_new':
/src/netmap/LINUX/../sys/dev/netmap/netmap_kern.h:261:18: error: storage size of '__xxts' isn't known
   struct timeval __xxts;    \
                  ^~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c:1759:3: note: in expansion of macro 'nm_prinf'
   nm_prinf("req if %d*%d ring %d*%d buf %d*%d",
   ^~~~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_kern.h:261:18: error: unused variable '__xxts' [-Werror=unused-variable]
   struct timeval __xxts;    \
                  ^~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c:1759:3: note: in expansion of macro 'nm_prinf'
   nm_prinf("req if %d*%d ring %d*%d buf %d*%d",
   ^~~~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c: In function 'netmap_free_rings':
/src/netmap/LINUX/../sys/dev/netmap/netmap_kern.h:261:18: error: storage size of '__xxts' isn't known
   struct timeval __xxts;    \
                  ^~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c:1870:6: note: in expansion of macro 'nm_prinf'
      nm_prinf("NOT deleting ring %s (ring %p, users %d neekring %d)",
      ^~~~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_kern.h:261:18: error: unused variable '__xxts' [-Werror=unused-variable]
   struct timeval __xxts;    \
                  ^~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c:1870:6: note: in expansion of macro 'nm_prinf'
      nm_prinf("NOT deleting ring %s (ring %p, users %d neekring %d)",
      ^~~~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_kern.h:261:18: error: storage size of '__xxts' isn't known
   struct timeval __xxts;    \
                  ^~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c:1875:5: note: in expansion of macro 'nm_prinf'
     nm_prinf("deleting ring %s", kring->name);
     ^~~~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_kern.h:261:18: error: unused variable '__xxts' [-Werror=unused-variable]
   struct timeval __xxts;    \
                  ^~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c:1875:5: note: in expansion of macro 'nm_prinf'
     nm_prinf("deleting ring %s", kring->name);
     ^~~~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c: In function 'netmap_mem2_rings_create':
/src/netmap/LINUX/../sys/dev/netmap/netmap_kern.h:261:18: error: storage size of '__xxts' isn't known
   struct timeval __xxts;    \
                  ^~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c:1911:6: note: in expansion of macro 'nm_prinf'
      nm_prinf("NOT creating ring %s (ring %p, users %d neekring %d)",
      ^~~~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_kern.h:261:18: error: unused variable '__xxts' [-Werror=unused-variable]
   struct timeval __xxts;    \
                  ^~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c:1911:6: note: in expansion of macro 'nm_prinf'
      nm_prinf("NOT creating ring %s (ring %p, users %d neekring %d)",
      ^~~~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_kern.h:261:18: error: storage size of '__xxts' isn't known
   struct timeval __xxts;    \
                  ^~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c:1916:5: note: in expansion of macro 'nm_prinf'
     nm_prinf("creating %s", kring->name);
     ^~~~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_kern.h:261:18: error: unused variable '__xxts' [-Werror=unused-variable]
   struct timeval __xxts;    \
                  ^~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c:1916:5: note: in expansion of macro 'nm_prinf'
     nm_prinf("creating %s", kring->name);
     ^~~~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_kern.h:270:18: error: storage size of '__xxts' isn't known
   struct timeval __xxts;    \
                  ^~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c:1922:5: note: in expansion of macro 'nm_prerr'
     nm_prerr("Cannot allocate %s_ring", nm_txrx2str(t));
     ^~~~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_kern.h:270:18: error: unused variable '__xxts' [-Werror=unused-variable]
   struct timeval __xxts;    \
                  ^~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c:1922:5: note: in expansion of macro 'nm_prerr'
     nm_prerr("Cannot allocate %s_ring", nm_txrx2str(t));
     ^~~~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_kern.h:261:18: error: storage size of '__xxts' isn't known
   struct timeval __xxts;    \
                  ^~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c:1945:6: note: in expansion of macro 'nm_prinf'
      nm_prinf("allocating buffers for %s", kring->name);
      ^~~~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_kern.h:261:18: error: unused variable '__xxts' [-Werror=unused-variable]
   struct timeval __xxts;    \
                  ^~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c:1945:6: note: in expansion of macro 'nm_prinf'
      nm_prinf("allocating buffers for %s", kring->name);
      ^~~~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_kern.h:270:18: error: storage size of '__xxts' isn't known
   struct timeval __xxts;    \
                  ^~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c:1947:6: note: in expansion of macro 'nm_prerr'
      nm_prerr("Cannot allocate buffers for %s_ring", nm_txrx2str(t));
      ^~~~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_kern.h:270:18: error: unused variable '__xxts' [-Werror=unused-variable]
   struct timeval __xxts;    \
                  ^~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c:1947:6: note: in expansion of macro 'nm_prerr'
      nm_prerr("Cannot allocate buffers for %s_ring", nm_txrx2str(t));
      ^~~~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_kern.h:261:18: error: storage size of '__xxts' isn't known
   struct timeval __xxts;    \
                  ^~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c:1953:6: note: in expansion of macro 'nm_prinf'
      nm_prinf("NOT allocating buffers for %s", kring->name);
      ^~~~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_kern.h:261:18: error: unused variable '__xxts' [-Werror=unused-variable]
   struct timeval __xxts;    \
                  ^~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c:1953:6: note: in expansion of macro 'nm_prinf'
      nm_prinf("NOT allocating buffers for %s", kring->name);
      ^~~~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c: In function 'netmap_mem2_deref':
/src/netmap/LINUX/../sys/dev/netmap/netmap_kern.h:261:18: error: storage size of '__xxts' isn't known
   struct timeval __xxts;    \
                  ^~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c:2075:3: note: in expansion of macro 'nm_prinf'
   nm_prinf("active = %d", nmd->active);
   ^~~~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_kern.h:261:18: error: unused variable '__xxts' [-Werror=unused-variable]
   struct timeval __xxts;    \
                  ^~~~~~
/src/netmap/LINUX/../sys/dev/netmap/netmap_mem2.c:2075:3: note: in expansion of macro 'nm_prinf'
   nm_prinf("active = %d", nmd->active);
   ^~~~~~~~
cc1: all warnings being treated as errors
make[3]: *** [/src/netmap/LINUX/Kbuild:23: /src/netmap/LINUX/netmap_mem2.o] Error 1
make[2]: *** [/src/linux/Makefile:1683: /src/netmap/LINUX] Error 2
make[1]: *** [/src/linux/Makefile:180: sub-make] Error 2
make[1]: Leaving directory '/src/linux-build'
make: *** [netmap.mak:81: netmap.ko] Error 2
```
</details>

--- a/LINUX/bsd_glue.h
+++ b/LINUX/bsd_glue.h
@@ -440,6 +440,13 @@ struct nm_linux_selrecord_t;
 
 #define	tsleep(a, b, c, t)	msleep(10)
 
+#ifndef NETMAP_LINUX_HAVE_STRUCT_TIMEVAL
+struct timeval {
+	long	tv_sec;		/* seconds */
+	long	tv_usec;	/* microseconds */
+};
+#endif /* !NETMAP_LINUX_HAVE_STRUCT_TIMEVAL */
+
 #define microtime		do_gettimeofday		/* debugging */
 #ifndef NETMAP_LINUX_HAVE_DO_GETTIMEOFDAY
 #define do_gettimeofday(tv_)					\

--- a/LINUX/configure
+++ b/LINUX/configure
@@ -1740,6 +1740,18 @@ EOF
 	}
 EOF
 
+  # is struct timeval defined?
+  add_test 'have STRUCT_TIMEVAL' << EOF
+	#include <linux/time.h>
+
+	void
+	dummy(struct timeval *tv) {
+		tv->tv_sec = 0;
+		tv->tv_usec = 0;
+		return;
+	}
+EOF
+
   add_test 'have DO_GETTIMEOFDAY' <<EOF
 	#include <linux/time.h>
 

